scarb/src/compiler/plugin/proc_macro/host.rs

```diff
@@ -16,22 +22,131 @@ pub struct ProcMacroHostPlugin {
     macros: Vec<Arc<ProcMacroInstance>>,
 }
 
+pub type ProcMacroId = SmolStr;
+
+#[derive(Debug)]
+#[allow(dead_code)]
+pub enum ProcMacroKind {
+    /// `proc_macro_name!(...)`
+    MacroCall,
+    /// `#[proc_macro_name]`
+    Attribute,
+    /// `#[derive(...)]`
+    Derive,
+}
+
+#[derive(Debug)]
+pub struct ProcMacroInput {
+    pub id: ProcMacroId,
+    pub kind: ProcMacroKind,
+}
+
 impl ProcMacroHostPlugin {
     pub fn new(macros: Vec<Arc<ProcMacroInstance>>) -> Self {
         Self { macros }
     }
+
+    /// Handle `proc_macro_name!` expression.
+    fn handle_macro(
+        &self,
+        _db: &dyn SyntaxGroup,
+        _item_ast: ast::ModuleItem,
+    ) -> Vec<ProcMacroInput> {
+        // Todo(maciektr): Implement.
+        Vec::new()
+    }
+
+    /// Handle `#[proc_macro_name]` attribute.
+    fn handle_attribute(
+        &self,
+        db: &dyn SyntaxGroup,
+        item_ast: ast::ModuleItem,
+    ) -> Vec<ProcMacroInput> {
+        let attrs = match item_ast {
+            ast::ModuleItem::Struct(struct_ast) => Some(struct_ast.attributes(db)),
+            ast::ModuleItem::Enum(enum_ast) => Some(enum_ast.attributes(db)),
+            ast::ModuleItem::ExternType(extern_type_ast) => Some(extern_type_ast.attributes(db)),
+            ast::ModuleItem::ExternFunction(extern_func_ast) => {
+                Some(extern_func_ast.attributes(db))
+            }
+            ast::ModuleItem::FreeFunction(free_func_ast) => Some(free_func_ast.attributes(db)),
+            _ => None,
+        };
+
+        attrs
+            .map(|attrs| attrs.structurize(db))
+            .unwrap_or_default()
+            .iter()
+            .filter(|attr| self.declared_attributes().contains(&attr.id.to_string()))
```

Comment:
Was looking at the code and I don't seem to get why `declared_attributes` for `ProcMacroInstance` return its package name - is it some simplification we want to do at early stages of implementation?

Reply:
@piotmag769  Yes, this stems from the one macro per package simplification. Ultimately, it should discover macros exposed by the shared library, and load all them with their names. Right now I've skipped that part, so I would not need to implement any so contents discovery yet.